sdk/storage/storage-file/samples/typescript/iterators-files-and-directories.ts

```diff
@@ -0,0 +1,181 @@
+  const pipeline = newPipeline(sharedKeyCredential);
+
+  // List shares
+  const serviceClient = new FileServiceClient(
+    // When using AnonymousCredential, following url should include a valid SAS
+    `https://${account}.file.core.windows.net`,
+    pipeline
+  );
+
+  console.log(`List shares`);
+
+  // Create a share
+  const shareName = `newshare${new Date().getTime()}`;
+  const shareClient = serviceClient.createShareClient(shareName);
+  await shareClient.create();
+  console.log(`Create share ${shareName} successfully`);
+
+  // Create a directory
+  const directoryName = `newdirectory${new Date().getTime()}`;
+  const directoryClient = shareClient.createDirectoryClient(directoryName);
+  await directoryClient.create();
+  console.log(`Create directory ${directoryName} successfully`);
+
+  const content = "Hello World!";
+  const fileName = "newfile" + new Date().getTime();
+
+  // Creates 3 files and 3 directories in the above directory
+  for (let i = 0; i < 3; i++) {
+    const directoryClient2 = directoryClient.createDirectoryClient(directoryName + "-sub-" + i);
+    await directoryClient2.create();
+    console.log(`Create sub directory ${directoryName + "-sub-" + i} successfully`);
+
+    const fileClient = directoryClient.createFileClient(fileName + "-sub-" + i);
+    await fileClient.create(content.length);
+    console.log(`Create file ${fileName + "-sub-" + i} successfully`);
+  }
+
+  console.log(`List files and directories under directory ${directoryName}`);
+
+  ////////////////////////////////////////////////////////
+  /////////////  List files and directories  /////////////
+  ////////////////////////////////////////////////////////
+
+  let i = 1;
+
+  let dirIter1 = directoryClient.listFilesAndDirectories();
+  for await (const item of dirIter1) {
+    if (item.kind === "directory") {
+      console.log(`${i} - directory\t: ${item.name}`);
+    } else {
+      console.log(`${i} - file\t: ${item.name}`);
+    }
+    i++;
+  }
+
+  // Same as the previous example
+  i = 1;
+  for await (const item of directoryClient.listFilesAndDirectories()) {
+    if (item.kind === "directory") {
+      console.log(`${i} - directory\t: ${item.name}`);
+    } else {
+      console.log(`${i} - file\t: ${item.name}`);
+    }
+    i++;
+  }
+
+  // Generator syntax .next()
+  let dirIter2 = await directoryClient.listFilesAndDirectories();
+  i = 1;
+  let item = await dirIter2.next();
+  while (!item.done) {
+    if (item.value.kind === "directory") {
+      console.log(`${i} - directory\t: ${item.value.name}`);
+    } else {
+      console.log(`${i} - file\t: ${item.value.name}`);
+    }
+    item = await dirIter2.next();
+    i++;
+  }
+
+  ///////////////  Examples for .byPage()  ///////////////
+
+  i = 1;
+  for await (const item1 of directoryClient.listFilesAndDirectories().byPage()) {
+    for (const fileItem of item1.segment.fileItems) {
+      console.log(`${i} - file\t: ${fileItem.name}`);
```

Comment:
**Concerns**
`byPage()` returns asynciterableiterator for segment responses.
The segment contains `fileitems` array and we can access it only by `item.segment.fileItems`.
```typescript
export interface FileItem {
  name: string;
  properties: FileProperty;
}
```
This is not in coherence with the async iterable iterator for `AzureFileItem` and `AzureDirectoryItem` .
```typescript
export interface AzureFileItem {
  kind: "file";
  name: string;
  properties: Models.FileProperty;
}
```

Reply:
This might be fine

1. When consuming `segment.fileItems` or `segment.directoryItems` it's clear what type the item has thus no confusion.
2. In code wherer `FileItem` is expected we could pass in an `AzureFileItem`.
3. There would be problem if `AzureFileItem` is expected. Users would either need to add the `kind`.  We may be able to mitigate this case if we inline the type instead of introducing new type `AzureFileItem`? 

```ts
async listDirectorysAndFiles(): Promise<
  | { kind: "file" } & FileItem
  | { kind: "directory" } & DirectoryItem }> {
}
```
This way it's clear to users that we are adding `kind` on top of `FileItem`.

Reply:
I like the anonymous types idea, but maybe more so because it means we don't need to have similarly-named but subtly different types floating around.